File: typescript/tests/validate.test.ts
import { describe, it, expect } from 'vitest';
import { validatePassport, validateDNA, ValidationError } from '../src/validate.js';

const HEX64 = '0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789';
const DID = 'did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH';

const validPassport = () => ({
  '@context': 'https://agentpassport.org/v0.1',
  spec_version: '0.1.0',
  type: 'AgentPassport',
  id: DID,
  keys: { signing: { algorithm: 'Ed25519', public_key: 'z6MkpTHR' } },
  genesis_owner: { id: DID, bound_at: '2025-01-01T00:00:00Z', immutable: true },
  current_owner: { id: DID },
  snapshot: {
    version: 1,
    hash: HEX64,
    prev_hash: null,
    created_at: '2025-01-01T00:00:00Z',
    skills: {
      entries: [{ name: 'go', version: '1.0', description: 'Go', capabilities: ['code'], hash: HEX64 }],
      frozen: false,
    },
    soul: { personality: 'f', work_style: 't', constraints: [], hash: HEX64, frozen: false },
    policies: { policy_set_hash: HEX64, summary: ['can_bid'] },
  },
  lineage: { kind: 'single', parents: [], generation: 0 },
  proof: {
    type: 'Ed25519Signature2020',
    created: '2025-01-01T00:00:00Z',
    verification_method: `${DID}#keys-1`,
    proof_purpose: 'assertionMethod',
    proof_value: 'zSIG',
  },
});

const validDna = () => ({
  '@context': 'https://agentpassport.org/v0.2/dna',
  type: 'AgentDNA',
  agent_id: DID,
  version: 1,
  skills: [],
  soul: { personality: 'f', work_style: 't', constraints: [] },
  policies: { policy_set_hash: HEX64, summary: ['x'] },
  dna_hash: HEX64,
  frozen: false,
});

describe('validatePassport', () => {
  it('accepts valid passport', () => {
    expect(() => validatePassport(validPassport())).not.toThrow();
  });

  it('rejects missing @context', () => {
    const d = validPassport();
    delete (d as any)['@context'];
    expect(() => validatePassport(d)).toThrow(ValidationError);
  });

  it('rejects missing id', () => {
    const d = validPassport();
    delete (d as any).id;
    expect(() => validatePassport(d)).toThrow(ValidationError);
  });

  it('rejects wrong type for spec_version', () => {
    const d = validPassport();
    (d as any).spec_version = 123;
    expect(() => validatePassport(d)).toThrow(ValidationError);
  });

  it('rejects extra fields', () => {
    const d = { ...validPassport(), extra: 'hello' };
    expect(() => validatePassport(d)).toThrow(ValidationError);
  });

  it('rejects skills.entries wrong type', () => {
    const d = validPassport();
    (d as any).snapshot.skills.entries = 'not-array';
    expect(() => validatePassport(d)).toThrow(ValidationError);
  });
});

describe('validateDNA', () => {
  it('accepts valid DNA', () => {
    expect(() => validateDNA(validDna())).not.toThrow();
  });

  it('rejects missing dna_hash', () => {
    const d = validDna();
    delete (d as any).dna_hash;
    expect(() => validateDNA(d)).toThrow(ValidationError);
  });
});

File: go/validate/validate_test.go
package validate

import (
	"encoding/json"
	"testing"
)

func validPassportMap() map[string]interface{} {
	return map[string]interface{}{
		"@context":     "https://agentpassport.org/v0.1",
		"spec_version": "0.1.0",
		"type":         "AgentPassport",
		"id":           "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		"keys": map[string]interface{}{
			"signing": map[string]interface{}{
				"algorithm":  "Ed25519",
				"public_key": "z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
			},
		},
		"genesis_owner": map[string]interface{}{
			"id":        "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
			"bound_at":  "2025-01-01T00:00:00Z",
			"immutable": true,
		},
		"current_owner": map[string]interface{}{
			"id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		},
		"snapshot": map[string]interface{}{
			"version":    1,
			"hash":       "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
			"prev_hash":  nil,
			"created_at": "2025-01-01T00:00:00Z",
			"skills": map[string]interface{}{
				"entries": []interface{}{
					map[string]interface{}{
						"name":         "go-dev",
						"version":      "1.0.0",
						"description":  "Go development",
						"capabilities": []interface{}{"code_write"},
						"hash":         "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
					},
				},
				"frozen": false,
			},
			"soul": map[string]interface{}{
				"personality":  "focused",
				"work_style":   "test-first",
				"constraints":  []interface{}{},
				"hash":         "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
				"frozen":       false,
			},
			"policies": map[string]interface{}{
				"policy_set_hash": "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
				"summary":         []interface{}{"can_bid"},
			},
		},
		"lineage": map[string]interface{}{
			"kind":       "single",
			"parents":    []interface{}{},
			"generation": 0,
		},
		"proof": map[string]interface{}{
			"type":                "Ed25519Signature2020",
			"created":             "2025-01-01T00:00:00Z",
			"verificationMethod": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH#keys-1",
			"proofPurpose":       "assertionMethod",
			"proofValue":         "zSIG",
		},
	}
}

func toJSON(v interface{}) []byte {
	b, _ := json.Marshal(v)
	return b
}

func TestValidPassport(t *testing.T) {
	if err := ValidatePassport(toJSON(validPassportMap())); err != nil {
		t.Fatalf("valid passport failed: %v", err)
	}
}

func TestMissingContext(t *testing.T) {
	m := validPassportMap()
	delete(m, "@context")
	if err := ValidatePassport(toJSON(m)); err == nil {
		t.Fatal("expected error for missing @context")
	}
}

func TestMissingID(t *testing.T) {
	m := validPassportMap()
	delete(m, "id")
	if err := ValidatePassport(toJSON(m)); err == nil {
		t.Fatal("expected error for missing id")
	}
}

func TestWrongTypeSpecVersion(t *testing.T) {
	m := validPassportMap()
	m["spec_version"] = 123
	if err := ValidatePassport(toJSON(m)); err == nil {
		t.Fatal("expected error for numeric spec_version")
	}
}

func TestExtraFields(t *testing.T) {
	// passport schema has additionalProperties: false, so extra fields should fail
	m := validPassportMap()
	m["extra_field"] = "hello"
	err := ValidatePassport(toJSON(m))
	if err == nil {
		t.Fatal("expected error for extra fields (additionalProperties: false)")
	}
}

func TestValidDNA(t *testing.T) {
	dna := map[string]interface{}{
		"@context": "https://agentpassport.org/v0.2/dna",
		"type":     "AgentDNA",
		"agent_id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		"version":  1,
		"skills":   []interface{}{},
		"soul": map[string]interface{}{
			"personality": "focused",
			"work_style":  "test-first",
			"constraints": []interface{}{},
		},
		"policies": map[string]interface{}{
			"policy_set_hash": "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
			"summary":         []interface{}{"can_bid"},
		},
		"dna_hash": "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
		"frozen":   false,
	}
	if err := ValidateDNA(toJSON(dna)); err != nil {
		t.Fatalf("valid DNA failed: %v", err)
	}
}

func TestDNAMissingHash(t *testing.T) {
	dna := map[string]interface{}{
		"@context": "https://agentpassport.org/v0.2/dna",
		"type":     "AgentDNA",
		"agent_id": "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH",
		"version":  1,
		"skills":   []interface{}{},
		"soul": map[string]interface{}{
			"personality": "focused",
			"work_style":  "test-first",
			"constraints": []interface{}{},
		},
		"policies": map[string]interface{}{
			"policy_set_hash": "0xabcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
			"summary":         []interface{}{"can_bid"},
		},
		"frozen": false,
	}
	if err := ValidateDNA(toJSON(dna)); err == nil {
		t.Fatal("expected error for missing dna_hash")
	}
}

func TestSkillsEntriesWrongType(t *testing.T) {
	m := validPassportMap()
	snap := m["snapshot"].(map[string]interface{})
	snap["skills"] = map[string]interface{}{
		"entries": "not-an-array",
		"frozen":  false,
	}
	if err := ValidatePassport(toJSON(m)); err == nil {
		t.Fatal("expected error for skills.entries wrong type")
	}
}
